[2단계 - 장기] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 제이미! 2단계 리뷰를 맡아주셔서 감사드립니다!🙇‍♂️
2단계 미션을 진행하면서 궁성 기능과 db연결 까지는 구현하였지만 CRUD기능을 구현하지 못한채로 제출하게 되었습니다.

리뷰 잘 부탁 드립니다 : )
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 기물의 이동 로직에서 분기 처리가 미흡한 부분이 있습니다.
- 인덴트가 2인 로직들이 있습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### 궁성에서의 움직임을 갖는 기물들과 궁성과 관련없는 기물들을 따로 추상화 하여 관리

2단계를 진행하면서 궁성의 움직임에 영향을 받는 기물과 그렇지 않은 기물들을 따로 관리해주기 위해서 각각의 추상 클래스를 도입하여 관리해주었습니다. 이렇게 한 결과 이러한 장점을 느낄 수 있었습니다.

- 궁성에 영향을 받지 않는 기물들은 기존 코드의 수정이 없이 사용가능
- 궁성을 기준으로 추상클래스를 나눴지만 하나의 최상위 부모 Piece 타입으로 모든 기물 관리 가능

### 궁성 영역 Enum vs Class

궁성 영역을 가지는 클래스를 구현하기 위해서 Enum방식과 Class 방식을 고민하다 Class 방식으로 구현했는데, 생각보다 해당 클래스의 책임이 적은것 같은 느낌이 들었습니다. 이런 상황에서 어떤 기준으로 Enum과  Class를 사용해야 하는지 판단이 서지 않았습니다.
이에 대한 제이미의 생각이 궁금합니다!

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - 게임 규칙이 업데이트되어 대각선 및 한 칸 이동 방식이 추가되고, 궁성 영역(궁궐) 내에서의 기물 이동 규칙이 도입되었습니다.
  - 각 기물에 점수가 부여되어 팀별 총 점수 계산 및 탈락 기물에 따른 감점 기능이 새롭게 적용되었습니다.
  - 게임 종료 시 승리 팀과 최신 스코어를 명확하게 표시하는 기능이 개선되었습니다.

- **Refactor**
  - 게임 진행, 이동 검증 및 점수 계산 로직이 모듈화되어 전체 게임 흐름이 명확하고 유지보수하기 쉽게 개선되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dye0p committed:
docs: 2단계 기능 요구 사항 업데이트

dye0p committed:
feat: 궁성 객체 추가

dye0p committed:
refactor: 궁성 초기화 객체 네이밍 오타 수정

dye0p committed:
refactor: 궁성의 영향을 받는 기물과 받지 않는 기물을 각각의 추상 클래스로 분리

dye0p committed:
feat: 궁성에서의 왕 움직임 추가

dye0p committed:
feat: 궁성에서의 사 움직임 추가

dye0p committed:
docs: 기능 요구사항 수정

dye0p committed:
refactor: 뒤를 판단하는 로직 변경

dye0p committed:
feat: 한칸 또는 한칸 이상의 대각선을 판단한다.

dye0p committed:
refactor: 궁성 포함 확인 로직 스트림 제거

dye0p committed:
refactor: 궁성 초기화 로직 영역에 구분 개행 추가

dye0p committed:
feat: 궁성에서의 졸/병 움직임 추가

dye0p committed:
refactor: 졸/병 생성시 팀 고정으로 변경

- 졸/병은 팀이 이미 정해져있기 때문에 외부에서 팀을 변경하지 못하도록 변경하였음.

dye0p committed:
feat: 궁성에서의 포 움직임 추가

dye0p committed:
feat: 궁성에서의 포 움직임 추가

dye0p committed:
feat: 궁성에서의 차 움직임 추가

dye0p committed:
docs: 기능 요구 사항 완료 사항 업데이트

dye0p committed:
fix: 사 기물 기본 배치 올바른 위치로 변경

dye0p committed:
feat: 라운드마다 점수를 계산하고, 기물이 탈락되면 점수에 반영한다.

dye0p committed:
refactor: 궁성 관련 클래스 궁성 패키지로 이동

dye0p committed:
refactor: 관련된 메서드 메서드 분리

dye0p committed:
chore: mysql-connector 의존성 추가

dye0p committed:
feat: db 구성

- 초기 sql 구성
- db connection 확인 테스트

jamie9504 commented on .gitignore:
```diff
@@ -30,3 +30,4 @@ out/
 
```
> 궁성 영역 Enum vs Class
> 궁성 영역을 가지는 클래스를 구현하기 위해서 Enum방식과 Class 방식을 고민하다 Class 방식으로 구현했는데, 생각보다 해당 클래스의 책임이 적은것 같은 느낌이 들었습니다. 이런 상황에서 어떤 기준으로 Enum과 Class를 사용해야 하는지 판단이 서지 않았습니다.
이에 대한 제이미의 생각이 궁금합니다!

정해진 답은 없지만, 버퍼 기간이 있으므로 모두 구현한 후 브랜치를 새로 하나 만들어서 Enum 방식으로 구현해보고 느낀 장/단점을 비교해보면 좋겠네요.
이후 구체화한 장/단점으로 크루들과 관련하여 논의해보세요!

jamie9504 reviewed (changes requested):
안녕하세요. 도기, 리뷰어 제이미입니다.

프로그램 요구사항을 구현하지 못했다면 기한을 넘긴 것과 다름이 없기에 바로 Request Changes 보냅니다.

일정을 맞추지 못한 이유에 대해 충분히 고민하고, 앞으로는 같은 상황에서 어떻게 하면 좋을지 방안을 수립해보세요.

다음 번 요청은 마감기한이 지남에 따라 별도의 리뷰 없이 Approve & Merge 예정입니다.
수정하고 싶은 부분까지 충분히 수정해주세요.

dye0p committed:
chore: piece 테이블 컬럼명 수정 및 주석제거

dye0p committed:
docs: 데이터베이스 요구사항 추가

dye0p committed:
feat: 게임을 db에 생성하고 진행상태를 저장하는 기능 구현

- db에서 게임을 찾고, 없다면 새로운 게임을 생성한다.
- 피스의 정보를 db에 저장한다.
- 프로그램을 재시작 하더라도 이전의 게임을 계속해서 진행 할 수 있다.

dye0p committed:
feat: 게임이 종료되면 db에서 게임과 피스의 정보를 삭제한다.

dye0p committed:
refactor: 메서드 네이밍 변경

dye0p committed:
refactor: db 처리 로직 서비스 클래스로 분리

dye0p committed:
refactor: 각각의 dao 인터페이스를 구현하도록 변경

- dao가 인터페이스를 구현하도록 변경하여 service 클래스에서 인터페이스에 의존하도록 변경하였음.

dye0p committed:
test: JanggiService 테스트 추가

- dao의 fake 객체를 생성하여 테스트를 진행할 수 있다.

dye0p committed:
refactor: 메서드 분리 및 포멧팅

dye0p committed:
feat: 이전 게임, 새로운 게임의 상태에 따라서 안내 메세지를 출력하는 메세지 추가

dye0p committed:
refactor: dao 패키지 구분

dye0p committed:
refactor: 궁성 클래스 Enum으로 변경

- 궁성은 장기에서 고유의 영역으로 영역이 수정될 가능성이 적다고 판단하여 Enum으로 변경하여 상수화 하였음

dye0p committed:
chore: sql 생성 폴더 경로 변경

dye0p committed:
refactor: jdbc 커넥터 정적 메서드로 변경

dye0p commented:
안녕하세요 제이미! 추가적인 db 구현까지 완료하여 모든 필수 요구사항을 충족하였습니다!
레벨1 마지막 미션의 리뷰이로 만나게 되어서 영광이었습니다!🙇‍♂️ 수고하셨습니다 : )

## 프로그램 실행 방법
1. java-janggi/docker 폴더로 이동한다.
2. docker-compose -p janggi up -d 명령어를 실행하여 mySql 서버를 실행한다.
3. 도커에 janggi db가 실행되었다면 애플리케이션을 실행한다.

--- 

이번 2단계에서 마감기한을 지키기 못한 이유를 생각해보았습니다.

### 1. 구현 속도의 부족함

매번 미션을 진행하면서 느끼는 부분이었지만 구현 속도가 비교적 느린것 같습니다.

- 생각의 시간이 오래걸림
- 생각한 대로 코드구현의 어려움

이러한 단점을 보완하기 위해서 빠르게 구현하는 연습이 필요함을 느꼈습니다.

### 2. 요구사항 이해 부족

요구사항을 정확하게 숙지하지 않아 구현하는 과정에서 도메인 개념이 헷갈리는 경우가 많았습니다.
이러한 단점을 보완하기 위해서 초반 구현을 내려놓고 요구사항을 분석하고 설계하는 시간을 조금 더 길게 가져가야함을 느꼈습니다.


추가로 질문의 답변에 대한 생각을 정리해보았습니다.

## 궁성 영역을 Enum으로 가져가기

궁성의 영역을 구현하는 방식으로 class  또는 Enum을 고민하였고 최종적으로 기존의 Class를  Enum으로 변경하였습니다. 
변경하는 과정에서 둘의 장단점을 정리해보았습니다.

### Class

장점

- 클래스로 선언하게 되면 추후에 궁성의 영역이 추가/삭제 또는 영역의 변경에 대비할 수 있다.

단점

- 이전의 방식에서 `PalaceGenerator`를 이용하여 궁성의 위치를 생성하고, `addArea()`로 해당 영역을 Palace 클래스로 추가하였는데,
addArea() 메서드가 public으로 열려있기 때문에 외부에서 접근이 가능하다.

### Enum

장점

- 상수로 정의되어 있기때문에 기본적으로 불변 객체다.
- 항상 같은 인스턴스로 사용할 수 있다.
- 궁성 영역이라는 상수를 객체처럼 사용할 수 있다.

단점

- 상수로 선언되어있기 때문에 궁성의 영역을 런타임 시점에서 동적으로 설정할 수 없음

각각의 장단점이 있지만 현재 구조를 고려했을 때 궁성은 변경될 가능성이 적다고 판단되어 Enum으로 변경하였습니다!

dye0p committed:
test: db 연결 테스트 수정

dye0p commented:
안녕하세요 제이미! 추가적인 db 구현까지 완료하여 모든 필수 요구사항을 충족하였습니다!
레벨1 마지막 미션의 리뷰이로 만나게 되어서 영광이었습니다!🙇‍♂️ 수고하셨습니다 : )

## 프로그램 실행 방법
1. java-janggi/docker 폴더로 이동한다.
2. docker-compose -p janggi up -d 명령어를 실행하여 mySql 서버를 실행한다.
3. janggi db가 실행되었다면 애플리케이션을 실행한다.

--- 

이번 2단계에서 마감기한을 지키기 못한 이유를 생각해보았습니다.

### 1. 구현 속도의 부족함

매번 미션을 진행하면서 느끼는 부분이었지만 구현 속도가 비교적 느린것 같습니다.

- 생각의 시간이 오래걸림
- 생각한 대로 코드구현의 어려움

이러한 단점을 보완하기 위해서 빠르게 구현하는 연습이 필요함을 느꼈습니다.

### 2. 요구사항 이해 부족

요구사항을 정확하게 숙지하지 않아 구현하는 과정에서 도메인 개념이 헷갈리는 경우가 많았습니다.
이러한 단점을 보완하기 위해서 초반 구현을 내려놓고 요구사항을 분석하고 설계하는 시간을 조금 더 길게 가져가야함을 느꼈습니다.


추가로 질문의 답변에 대한 생각을 정리해보았습니다.

## 궁성 영역을 Enum으로 가져가기

궁성의 영역을 구현하는 방식으로 class  또는 Enum을 고민하였고 최종적으로 기존의 Class를  Enum으로 변경하였습니다. 
변경하는 과정에서 둘의 장단점을 정리해보았습니다.

### Class

장점

- 클래스로 선언하게 되면 추후에 궁성의 영역이 추가/삭제 또는 영역의 변경에 대비할 수 있다.

단점

- 이전의 방식에서 `PalaceGenerator`를 이용하여 궁성의 위치를 생성하고, `addArea()`로 해당 영역을 Palace 클래스로 추가하였는데,
addArea() 메서드가 public으로 열려있기 때문에 외부에서 접근이 가능하다.

### Enum

장점

- 상수로 정의되어 있기때문에 기본적으로 불변 객체다.
- 항상 같은 인스턴스로 사용할 수 있다.
- 궁성 영역이라는 상수를 객체처럼 사용할 수 있다.

단점

- 상수로 선언되어있기 때문에 궁성의 영역을 런타임 시점에서 동적으로 설정할 수 없음

각각의 장단점이 있지만 현재 구조를 고려했을 때 궁성은 변경될 가능성이 적다고 판단되어 Enum으로 변경하였습니다!

jamie9504 commented on .gitignore:
```diff
@@ -30,3 +30,4 @@ out/
 
 ### VS Code ###
```
이번 단계에서는 가장 중요한 `데드라인을 지키는 것`과 `지키기 위해 어떻게 해야 하는지`를 학습했네요.
아쉬울 순 있지만 일찍 알게 될수록 좋은 부분입니다.

---

구현 시, 고민되는 부분이 간단하다면 직접 구현해보는 것도 장/단점을 확실하게 알 수 있는 방법 중 하나입니다.
어떤 방식으로 장/단점을 구분하여 가장 적합하다고 생각하는 방식을 고를 수 있을 지 생각해보는 것도 좋겠네요.

jamie9504 reviewed (approved):
안녕하세요, 도기. 리뷰어 제이미입니다.

Approve & Merge 합니다.

1레벨 마지막 미션 고생 많으셨고, 버퍼기간 & 방학기간동안 충분히 재충전하셔서
남은 우테코 기간도 즐겁게 보내시기 바랍니다.
